aws/resource_aws_sns_platform_application.go

```diff
@@ -0,0 +1,272 @@
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("Error updating SNS platform application: %s", err)
+	}
+
+	return resourceAwsSnsPlatformApplicationRead(d, meta)
+}
+
+func resourceAwsSnsPlatformApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	// There is no SNS Describe/GetPlatformApplication to fetch attributes like name and platform
+	// We will use the ID, which should be a platform application ARN, to:
+	//  * Validate its an appropriate ARN on import
+	//  * Parse out the name and platform
+	platformApplicationArn, err := arn.Parse(d.Id())
+	if err != nil {
+		return fmt.Errorf(
+			"SNS Platform Application ID must be of the form "+
+				"arn:PARTITION:sns:REGION:ACCOUNTID:app/PLATFORM/NAME, "+
+				"was provided %q and received error: %s", platformApplicationArn.String(), err)
+	}
+
+	platformApplicationArnResourceParts := strings.Split(platformApplicationArn.Resource, "/")
+	if len(platformApplicationArnResourceParts) != 3 || platformApplicationArnResourceParts[0] != "app" {
+		return fmt.Errorf(
+			"SNS Platform Application ID must be of the form "+
+				"arn:PARTITION:sns:REGION:ACCOUNTID:app/PLATFORM/NAME, "+
+				"was provided: %s", platformApplicationArn.String())
+	}
+
+	d.Set("arn", platformApplicationArn.String())
+	d.Set("name", platformApplicationArnResourceParts[2])
+	d.Set("platform", platformApplicationArnResourceParts[1])
+
+	attributeOutput, err := snsconn.GetPlatformApplicationAttributes(&sns.GetPlatformApplicationAttributesInput{
+		PlatformApplicationArn: aws.String(platformApplicationArn.String()),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if attributeOutput.Attributes != nil && len(attributeOutput.Attributes) > 0 {
+		attrmap := attributeOutput.Attributes
+		resource := *resourceAwsSnsPlatformApplication()
+		// iKey = internal struct key, oKey = AWS Attribute Map key
+		for iKey, oKey := range snsPlatformApplicationAttributeMap {
+			log.Printf("[DEBUG] Updating %s => %s", iKey, oKey)
+
+			if attrmap[oKey] != nil {
+				// Some of the fetched attributes are stateful properties such as
+				// the number of subscriptions, the owner, etc. skip those
+				if resource.Schema[iKey] != nil {
+					value := *attrmap[oKey]
+					log.Printf("[DEBUG] Updating %s => %s -> %s", iKey, oKey, value)
+					d.Set(iKey, *attrmap[oKey])
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func resourceAwsSnsPlatformApplicationDelete(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	log.Printf("[DEBUG] SNS Delete Application: %s", d.Id())
+	_, err := snsconn.DeletePlatformApplication(&sns.DeletePlatformApplicationInput{
+		PlatformApplicationArn: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func hashSum(contents interface{}) string {
+	return fmt.Sprintf("%x", sha256.Sum256([]byte(contents.(string))))
+}
```

Comment:
Nitpick: Do you mind moving this function into the test file as it's only used there?

Reply:
See `StateFunc: hashSum,` above in the attributes. If you look in #1101 there are some comments asking about this. We don't do this anywhere else that I know off the top of my head for storing obfuscated attribute values (to keep secrets out of the state). I don't have a strong opinion either which way on keeping it or not. It does make the acceptance testing harder when working with actual files rather than strings.

Reply:
Ah, doh 🤦‍♂️ , not sure why I didn't notice that.

Regarding hashing sensitive data - I don't want to set precedent either way, but I'm slightly more inclined to keeping it plain-text and adding a warning we have in docs of other similarly affected resources, like these:

https://www.terraform.io/docs/providers/aws/r/rds_cluster.html

instead of giving people false hope or setting expectation that it's ok to keep plain-text state file anywhere.